File: 00757. Set Intersection Size At Least Two/solution.cs
/**
 * Problem: 757. Set Intersection Size At Least Two
 *
 * Difficulty: Hard
 *
 * Language: C#
 *
 * Performance: Runtime - 4 ms (Beats 100%)
 */

public class Solution
{
  public int IntersectionSizeTwo(int[][] intervals)
  {
    // Sort the intervals based on their end points in ascending order.
    // If two intervals have the same end point, sort by their start points in descending order.
    Array.Sort(intervals, (a, b) => a[1] != b[1] ? a[1] - b[1] : b[0] - a[0]);

    // Initialize the count of elements in the intersection set to 0.
    int count = 0;

    // Initialize two variables to track the last two elements added to the set.
    // These are initially set to -1 to indicate no elements have been added yet.
    int lastElement = -1, secondLastElement = -1;

    // Iterate through each interval in the sorted list.
    foreach (var interval in intervals)
    {
      // Extract the start and end points of the current interval.
      int start = interval[0], end = interval[1];

      // Check if the current interval starts after the last element in the set.
      if (start > lastElement)
      {
        // Add the last two elements of the interval to the set.
        secondLastElement = end - 1;
        lastElement = end;

        // Increment the count by 2 since two elements are added.
        count += 2;
      }
      // Check if the current interval starts after the second last element in the set.
      else if (start > secondLastElement)
      {
        // Add the last element of the interval to the set.
        secondLastElement = lastElement;
        lastElement = end;

        // Increment the count by 1 since one element is added.
        count++;
      }
    }

    // Return the total count of elements in the intersection set.
    return count;
  }
}

File: 00757. Set Intersection Size At Least Two/solution.js
/**
 * Problem: 757. Set Intersection Size At Least Two
 *
 * Difficulty: Hard
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 6 ms (Beats 100%)
 */

/**
 * Find the minimum size of a set covering all intervals
 *
 * @param {number[][]} intervals - List of intervals
 *
 * @returns {number} - Minimum set size
 */
const intersectionSizeTwo = (intervals) => {
  // Sort intervals by their end points in ascending order.
  // If two intervals have the same end point, sort by start point in descending order.
  intervals.sort((a, b) => (a[1] !== b[1] ? a[1] - b[1] : b[0] - a[0]))

  // Initialize the result variable to store the minimum set size.
  let result = 0

  // Initialize the last array to keep track of the last two points in the set.
  let last = []

  // Iterate through each interval in the sorted intervals.
  intervals.forEach((interval) => {
    // Destructure the start and end points of the current interval.
    const [start, end] = interval

    // Check if the last array is empty or the second point in last is less than the start.
    if (last.length === 0 || last[1] < start) {
      // Add two points to the result as the interval is not covered.
      result += 2
      // Update the last array with the two largest points in the interval.
      last = [end - 1, end]
    }
    // Check if the first point in last is less than the start.
    else if (last[0] < start) {
      // Add one point to the result as only one more point is needed.
      result += 1
      // Update the last array to include the second point in last and the end point.
      last = [last[1], end]
    }
  })

  // Return the result as the minimum set size.
  return result
}
